### PR TITLE
Pipes -- Scrabble -- Kimberley and Diane

### PR DESCRIPTION
**Pull request by lenertzdiane**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | We didn't want or need to instantiate a new instance of 'scoring', we just want to use a behavior related to scoring. This way we can call it without creating a 'scoring', which wouldn't even make sense.    |
|   Describe an example of a test you wrote for a _nominal case_. | tests to check if words under seven letters were scored correctly.   	|
|   Describe an example of a test you wrote for an _edge case_. |  We checked if a six letter and seven letter word had equal scores that the seven letter word would win the tie, whether the seven letter word came before or after the six letter word in the array_of_words.	|
|   What was your final test coverage percentage?	|  100% 	|
|   Is there a specific piece of code you'd like feedback on?	|   highest_score_from in the Scoring class. We tried to refactor but had a hard time making it any more readable/concise	|

### CODE REVIEW

File: lib/Scoring.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+require 'pry'
+
+module Scrabble
+  class Scoring
+    attr_reader :word_score
+    @scoring_hash = { A: 1, E: 1, I: 1, O: 1, U: 1, L: 1, N: 1,
+      R: 1, S: 1, T: 1, D: 2, G: 2, B: 3, C: 3,
+      M: 3, P:3, F: 4, H: 4, V: 4, W: 4, Y: 4, K: 5,
+      J: 8, X: 8, Q: 10, Z: 10
```
Comment (droberts-sea) [inline lib/Scoring.rb:9]: Why not use a constant for this?

File: lib/Scoring.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      array_of_words.each do |word|
+        word_score = self.score(word)
+        word = word.upcase
+        if word_score > highest_score
+          highest_score = word_score
+          highest_score_word = word
+        elsif word_score == highest_score
+          if word.length < highest_score_word.length
+            unless highest_score_word.length == 7
+              highest_score_word = word
```
Comment (droberts-sea) [inline lib/Scoring.rb:42]: This tie-breaking logic gets pretty complex! It might be wise to break it out into a separate method, both to clean up this code and so that the logic can be tested independently.

File: Specs/wave-1-specs.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      Scrabble::Scoring.score("CAT").must_equal Scrabble::Scoring.score("cAt")
+    end
+
+    it "returns the correct score for the input" do
+      Scrabble::Scoring.score("cat").must_equal 5
+    end
+
+    it "returns the correct score for 7-letter words" do
+      Scrabble::Scoring.score("lizards").must_equal (17 + 50)
+    end
```
Comment (droberts-sea) [inline Specs/wave-1-specs.rb:30]: I like that you've split up the 17 and the 50 here. It's a small thing, but it really helps readability.

File: Specs/wave-1-specs.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+  describe "self.highest_score_from(array_of_words)" do
+
+    it "Scoring responds to self.highest_score_from" do
+      Scrabble::Scoring.must_respond_to :highest_score_from
+    end
+
+    it "returns correct highest score word from array of 2 words" do
+      word_array = ["cat", "apples", "zygote"]
+      Scrabble::Scoring.highest_score_from(word_array).must_equal "ZYGOTE"
```
Comment (droberts-sea) [inline Specs/wave-1-specs.rb:47]: This test claims to test an array of 2 words, but this array has 3 words.

File: lib/Player.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        @plays << word
+        @total_score += Scoring.score(word)
+      else
+        return false
+      end
+    end
+
+    def highest_scoring_word
+      @highest_scoring_word = Scoring.highest_score_from(@plays)
+    end
```
Comment (droberts-sea) [inline lib/Player.rb:27]: This method's logic is correct. However, you have an `attr_reader` for this instance variable above, which is overridden here. This is a little confusing. I wonder if you even need to store `highest_scoring_word` in an instance variable.

Comment (droberts-sea) [reply lib/Player.rb:27]: If you were to remove the instance variable, then `highest_word_score` below would have to be:

```ruby
def highest_word_score
  word = highest_scoring_word
  return Scoring.score(word)
end
```

File: lib/Player.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        @tiles << tile
+      end
+    end
+
+    private
+
+    def won?
+      if @total_score > 100
+        return true
+      else
```
Comment (droberts-sea) [inline lib/Player.rb:45]: Since `<` always returns either `true` or `false` already, this could be shortened to:

```ruby
def won?
  return @total_score > 100
end
```

File: lib/TileBag.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+      def draw_tiles(num)
+        num = num.to_i
+        drawn_tiles = []
+
+        num.times do
+          drawn_tiles << @bag_of_tiles.shuffle![0]
+          @bag_of_tiles.delete_at(0)
+        end
```
Comment (droberts-sea) [inline lib/TileBag.rb:23]: Shuffle changes the order of the entire bag, so you could get away with just doing it once, in `initialize`.

File: Specs/wave-3-specs.rb [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      bag_two_tiles_drawn = @new_tile_bag2.draw_tiles(7)
+      bag_two_tiles_drawn.wont_equal bag_one_tiles_drawn  ## I think this test wont work, and we'll need a do loop instead
+    end
+    #Check that it removed the tiles from the tile bag
+    it "must remove the tiles from the tile bag array once drawn" do
+      starting_length = @new_tile_bag.bag_of_tiles.length
+      @new_tile_bag.draw_tiles(3)
+      @new_tile_bag.bag_of_tiles.length.must_equal (starting_length - 3)
+    end
+
```
Comment (droberts-sea) [inline Specs/wave-3-specs.rb:61]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw the last tile from the bag?
- What happens when you attempt to draw more tiles than are in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  yes  |
|   Regular Commits with meaningful commit messages  |  yes  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  yes - empty string instead of 1-char string is OK  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `highest_scoring_word` edge cases (ties, empty array)	|  some - Good work testing for ties, I especially like that you swap the order and make sure it behaves the same. I would like to see a test for what happens when the input array is empty.  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  good coverage for when the player has already won, missing coverage around invalid words  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  yes  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes  |
| `tiles_remaining` method returns a count	|  no - this method is missing!  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  some - see inline comment  |

I have a few things to say regarding your project organization:
1. Files inside the lib folder should be named using snake case, that is, `tile_bag.rb` rather than `TileBag.rb`
1. Your `Specs` folder is spelled with a capital `S`, while the convention is lowercase
1. Your individual spec files should be named after the class they test. So `scoring_spec.rb` or `tile_bag_spec.rb`
1. The `coverage` folder generated by simplecov should be git-ignored

Other than that, this project looks good. I've identified a few places below where I noticed you're missing test coverage or where I thought something could be clarified, but I don't see any major issues. Keep up the good work!
